src/main/java/org/apache/commons/lang3/reflect/MethodUtils.java

```diff
@@ -0,0 +1,1024 @@
+     * can be found, return {@code null}.</p>
+     *
+     * <p>There isn't any good reason why this method must be {@code private}.
+     * It is because there doesn't seem any reason why other classes should
+     * call this rather than the higher level methods.</p>
+     *
+     * @param cls Parent class for the interfaces to be checked
+     * @param methodName Method name of the method we wish to call
+     * @param parameterTypes The parameter type signatures
+     * @return the accessible method or {@code null} if not found
+     */
+    private static Method getAccessibleMethodFromInterfaceNest(Class<?> cls,
+            final String methodName, final Class<?>... parameterTypes) {
+        // Search up the superclass chain
+        for (; cls != null; cls = cls.getSuperclass()) {
+
+            // Check the implemented interfaces of the parent class
+            final Class<?>[] interfaces = cls.getInterfaces();
+            for (final Class<?> anInterface : interfaces) {
+                // Is this interface public?
+                if (!ClassUtils.isPublic(anInterface)) {
+                    continue;
+                }
+                // Does the method exist on this interface?
+                try {
+                    return anInterface.getDeclaredMethod(methodName,
+                            parameterTypes);
+                } catch (final NoSuchMethodException e) { // NOPMD
+                    /*
+                     * Swallow, if no method is found after the loop then this
+                     * method returns null.
+                     */
+                }
+                // Recursively check our parent interfaces
+                final Method method = getAccessibleMethodFromInterfaceNest(anInterface,
+                        methodName, parameterTypes);
+                if (method != null) {
+                    return method;
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * <p>Finds an accessible method that matches the given name and has compatible parameters.
+     * Compatible parameters mean that every method parameter is assignable from
+     * the given parameters.
+     * In other words, it finds a method with the given name
+     * that will take the parameters given.</p>
+     *
+     * <p>This method is used by
+     * {@link
+     * #invokeMethod(Object object, String methodName, Object[] args, Class[] parameterTypes)}.
+     * </p>
+     *
+     * <p>This method can match primitive parameter by passing in wrapper classes.
+     * For example, a {@code Boolean} will match a primitive {@code boolean}
+     * parameter.
+     * </p>
+     *
+     * @param cls find method in this class
+     * @param methodName find method with this name
+     * @param parameterTypes find method with most compatible parameters
+     * @return The accessible method
+     */
+    public static Method getMatchingAccessibleMethod(final Class<?> cls,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 709-722 into a separate method.